service/s3/s3crypto/kms_key_provider.go

```diff
@@ -0,0 +1,127 @@
+package s3crypto
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/kms"
+)
+
+// KMSKeyProvider will make calls to KMS to get the masterkey
+type KMSKeyProvider struct {
+	kms          *kms.KMS
+	key          []byte
+	encryptedKey []byte
+	iv           []byte
+	cmkID        *string
+
+	CipherData
+	MaterialDescription
+}
+
+// NewKMSKeyProvider builds a new KMS key provider using the customer key ID and material
+// description.
+//
+// Example:
+// sess := session.New(&aws.Config{})
+// cmkID := "arn to key"
+// matdesc := s3crypto.NewJSONMatDesc()
+// kp, err := s3crypto.NewKMSKeyProvider(sess, cmkID, matdesc)
+func NewKMSKeyProvider(sess *session.Session, cmkID string, matdesc MaterialDescription) (KeyProvider, error) {
+	if matdesc == nil {
+		matdesc = &JSONMatDesc{}
+	}
+	matdesc.Set("kms_cmk_id", cmkID)
+
+	kp := &KMSKeyProvider{
+		MaterialDescription: matdesc,
+		kms:                 kms.New(sess),
+		cmkID:               &cmkID,
+	}
+	kp.Algorithm = "kms"
+	return kp, nil
+}
+
+// NewKMSKeyProviderWithMatDesc initializes a KMS keyprovider with a material description. This
+// is used with Decrypting kms content, due to the cmkID being in the material description.
+func NewKMSKeyProviderWithMatDesc(sess *session.Session, matdesc string) (KeyProvider, error) {
```

Comment:
Yea, I can definitely see the confusion there. The difference here is that the material description is populated with needed data. The other one is populated with whatever the user wants.